### PR TITLE
[feature] Add Licenses to Preprints, Allow License Updates via API [PREP-252]

### PR DESCRIPTION
**Pull request by caseyrollins**
- continuation of @hmoco's work in #6561 

#### Purpose
- Adds the ability to add and update preprint and node licenses via the API.

#### Changes
- **Makes node licenses writeable.**
- Adds `required_properties` field to the `LicenseSerializer`.
- Adds the `license` relationship field to the `PreprintSerializer` to represent the preprint/license relationship.
- Adds the `license_record` field to the `PreprintSerializer` to represent the preprint license record (license year and list of copyright holders).
- Adds `set_license()` to `website/project/licenses/` to DRY up the logic between `set_node_license()` and `set_preprint_license()`.
  - All of the necessary license validation is done in this function. 
- Adds preprint specific logs for license updates.
- Adds all of the tests for adding and updating a preprint license.

#### Side effects
- Node licenses are also now writeable via the API.

#### QA Notes
- Only admins should be able to succesfully update a node/preprint license.
- When adding/updating a preprint license, the chosen license must be in the list of acceptable licenses for the given preprint provider, or else an error will be thrown ("Invalid license chosen for <preprint_provider>").
- Non-four-digit years and years in the future are invalid license years. 
- If the chosen license requires additional metadata (i.e. license year or list of copyright holders), that information must be included in the request to successfully add/update the license.

#### Example Request Payloads
- `/v2/preprints/<preprint_id>/`, to add/update a simple license (year and copyright holders not required, e.g. CC0)
```
    "data": {
        "id": <preprint_id>,
        "type": "preprints",
        "relationships": {
            "license": {
                "data": {
                    "type": "licenses",
                    "id": <license_id>
                }
            }
        }
    }
```

- `/v2/preprints/<preprint_id>/`, to add/update a license where year and copyright holders must be specified, e.g. "No license"
```
    "data": {
        "id": <preprint_id>,
        "type": "preprints",
        "attributes": {
            "license_record": {
                "year": <year>,
                "copyright_holders": [<name>, <name>, <name>, ...]
            }
        }
        "relationships": {
            "license": {
                "data": {
                    "type": "licenses",
                    "id": <license_id>
                }
            }
        }
    }
```

#### Ticket
- [PREP-252](https://openscience.atlassian.net/browse/PREP-252)

### CODE REVIEW

File: api/licenses/serializers.py [hunk 1]
```diff
@@ -16,6 +16,7 @@ class LicenseSerializer(JSONAPISerializer):
     type = TypeField()
     name = ser.CharField(required=True, help_text='License name')
     text = ser.CharField(required=True, help_text='Full text of the license')
+    required_properties = ser.ListField(source='properties', read_only=True, help_text='Required attributes for the license')
```
Comment (brianjgeiger) [inline api/licenses/serializers.py]: I think `required_fields` would be better. `help_text='Fields required for this license to help front-end validators'` or similar.

**Comment by brianjgeiger** [comment]
Oh, and it _should_ be safe to point this at develop now. Everything should be merged in.

**Comment by brianjgeiger** [comment]
Merge conflict. Also, should the request payloads require `type` fields?

**Comment by caseyrollins** [comment]
@brianjgeiger, it doesn't appear that preprint updates currently require a `type` field.
https://github.com/CenterForOpenScience/osf.io/blob/develop/api_tests/preprints/views/test_preprint_detail.py#L14

**Pull request merged**